src/Castle.Facilities.AspNet.Mvc/AspNetMvcFacility.cs

```diff
@@ -0,0 +1,112 @@
+// Copyright 2004-2017 Castle Project - http://www.castleproject.org/
+// 
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+// 
+//     http://www.apache.org/licenses/LICENSE-2.0
+// 
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+using System.Reflection;
+using System.Web;
+using Castle.MicroKernel.Lifestyle;
+
+namespace Castle.Facilities.AspNet.Mvc
+{
+	using System;
+	using System.Web.Mvc;
+
+	using Castle.Facilities.AspNet.Mvc.Exceptions;
+	using Castle.MicroKernel;
+	using Castle.MicroKernel.Facilities;
+
+	public class AspNetMvcFacility : AbstractFacility
+	{
+		private bool autoCreateScopes;
+		private bool initialised = false;
+		private Assembly controllerAssembly;
+		private AspNetMvcControllerFactory controllerFactory;
+		private Action<IKernel, Type> beforeControllerResolved;
+		private Action<IKernel, object> afterControllerReleased;
+		private const string ContextScopeKey = "castle-windsor-facility-aspnet-mvc-scope";
+
+		public AspNetMvcControllerFactory ControllerFactory => controllerFactory;
+
+		protected override void Init()
+		{
+			ThrowIfControllerAssemblyWasNotSet();
+
+			controllerFactory = new AspNetMvcControllerFactory(this.Kernel);
+			controllerFactory.AddControllerAssembly(controllerAssembly);
+			ControllerBuilder.Current.SetControllerFactory(controllerFactory);
+
+			ApplyAutoLifestyleScoping();
+			SubscribeBeforeControllerCreatedEvent();
+			SubscribeAfterControllerReleaseEvent();
+
+			initialised = true;
+		}
+
+		public AspNetMvcFacility AddControllerAssembly<T>()
+		{
+			controllerAssembly = typeof(T).Assembly;
+			return this;
+		}
+
+		public AspNetMvcFacility BeforeControllerResolved(Action<IKernel, Type> beforeResolvedCallback)
+		{
+			this.beforeControllerResolved = beforeResolvedCallback;
+			if (initialised) SubscribeBeforeControllerCreatedEvent();
+			return this;
+		}
+
+		public AspNetMvcFacility AfterControllerReleased(Action<IKernel, object> afterReleasedCallback)
+		{
+			this.afterControllerReleased = afterReleasedCallback;
+			if (initialised) SubscribeAfterControllerReleaseEvent();
+			return this;
+		}
+
+		public AspNetMvcFacility WithLifestyleScopedPerWebRequest()
+		{
+			this.autoCreateScopes = true;
+			if (initialised) ApplyAutoLifestyleScoping();
+			return this;
+		}
+
+		private void ApplyAutoLifestyleScoping()
+		{
+			if (autoCreateScopes)
+			{
+				controllerFactory.BeforeControllerResolved += (k, t) =>
+					HttpContext.Current.Items.Add(ContextScopeKey, k.BeginScope());
```

Comment:
Does ASP.NET MVC use the same `HttpContext` from `System.Web` and it is just ASP.NET Web API that created another set of those abstractions? Couldn't we just use both facilities for ASP.NET MVC. If not, how were people using Windsor with MVC until now?

Reply:
> Does ASP.NET MVC use the same HttpContext from System.Web and it is just ASP.NET Web API that created another set of those abstractions? 

Yes. WebAPI was always a different animal which was PITA. In a webhost scenario it would use the same one but selfhost came along and wrecked that idea. 

> If not, how were people using Windsor with MVC until now?

They would google "Windsor MVC", find a controller factory and then copy it into their web project. Why I started writing the mvc fluentwindsor extension. I was tired of this approach.